sarc/core/scraping/users.py

```diff
@@ -0,0 +1,187 @@
+
+    # Add the matching ids of the new usermatch to make sure that we have all
+    # the ids that this user is known under.
+    assert (
+        name_dict.get(update.matching_id.name, update.matching_id) == update.matching_id
+    )
+    value.known_matches.add(update.matching_id)
+    for mid in update.known_matches:
+        assert name_dict.get(mid.name, mid) == mid
+        value.known_matches.add(mid)
+
+    for domain, credentials in update.associated_accounts.items():
+        if domain not in value.associated_accounts:
+            value.associated_accounts[domain] = credentials
+        else:
+            value.associated_accounts[domain].merge_with(credentials, truncate=True)
+
+    value.supervisor.merge_with(update.supervisor, truncate=True)
+    value.co_supervisors.merge_with(update.co_supervisors, truncate=True)
+
+    value.github_username.merge_with(update.github_username, truncate=True)
+    value.google_scholar_profile.merge_with(
+        update.google_scholar_profile, truncate=True
+    )
+
+
+def scrape_users(
+    scrapers: list[tuple[str, Any]], _cache_policy: CachePolicy = CachePolicy.use
+) -> Iterable[UserMatch]:
+    """
+    Perform user scraping and matching according to the list of plugins passed in.
+
+    The first plugin to specify information wins in case of conflict.
+
+    This returns one UserMatch structure per scraped user, across all plugins.
+    The collected information is aggregated amongst plugins, but not with the
+    information in the database.
+    """
+    raw_data: dict[str, tuple[bytes, Any]] = {}
+    for scraper_name, config_data in scrapers:
+        try:
+            scraper = get_user_scraper(scraper_name)
+        except KeyError as e:
+            raise ValueError("Invalid user scraper") from e
+        config = scraper.validate_config(config_data)
+        raw_data[scraper_name] = (scraper.get_user_data(config), config)
+
+    # TODO: save the raw data for cache purposes
+    #  - Should refactor cache so that we don't *have* to use the function wrapper interface
+
+    # UserMatches, referenced by plugin name and matching id
+    user_refs: dict[MatchID, UserMatch] = {}
+    for scraper_name, (rdata, config) in raw_data.items():
+        for userm in scraper.parse_user_data(config, rdata):
+            userm.matching_id.name = scraper_name
+            # First, get all the userm that match with this one.
+            prev_userms: list[UserMatch] = [userm]
+            prev = user_refs.get(userm.matching_id, None)
+            if prev is not None:
+                prev_userms.append(prev)
+            for mid in userm.known_matches:
+                prev = user_refs.get(mid, None)
+                if prev is not None:
+                    prev_userms.append(prev)
+            # Second, filter out duplicates and sort the rest according to plugin rank
+            scraper_names = [name for name, _ in scrapers]
+            matching_userms = sorted(
+                set(prev_userms),
```

Comment:
Since user matches are hashed only with matching_id, `set(prev_userms)` will remove some duplicates which do have same matching_id but may have different values in other fields. Is it what we want ?

I think we should either fully hash user match, or only sort prev_userms without trying to remove duplicates: if there are indeed full duplicates, they will be merged without changes and just few more time spent, I guess ?

Reply:
Depending on the configuration, it could indeed be the case that a plugin is called more than once with different configurations and thus could return different data for the same ID.

The problem I have with your solution is that in the normal case (no MatchID collision) there will be a lot of duplicate instances of the same result in the list of candidates and thus we will perform a lot of extra work to merge them all, since we replace the merged entries with the result of the merge.

Also, fully hashing the structure is a lot of extra work in `__hash__` and `__eq__` since the structure contains lots of sub-fields.

I'm not sure what would be the best solution here to fix that problem, but I would like to avoid adding a lot of extra work to the normal path if we can.